GoogleForCreators/web-stories-wp
# Dashboard infinite scroll implementation (My Stories)

# Overview
Gets my stories dashboard view hooked up to infinite scroll. The biggie here is that it converts the story state in the context provider to a reducer so that we can story pages of stories and track the order they are supposed to be presented. 

## Updates to context 
- we now have reducer/stories, where stories related state can be tracked. 
- I separated out a stories object and an array of ids that correspond to stories returned. the array will track the order of what we want displayed since order is tied to the results. The stories are re-sorted by the view with the array of ids and then used in the same format we're used to. This is important! 
- we need to update the array of items for what comes back to only be unique ids so that stories don't display more than once. This is also why you'll see that the stories and the order (and totals) get cleared out when filtering/searching happen. 
- Added a loading state. isLoading is true as soon as `fetchStories` gets called, then added a `.finally` to put `isLoading` back to false. 

## views/myStories/index.js 
- Few things I want to call out here - sorting and searching 
- we have to keep clearing out the existing order and stories object with search because the results might be different. i debounced it, but if you have other ideas i'm all ears. 
- something similar with sort really, we start from the top when new order is selected so that pagination has a chance to reset. 
- ~There is a bug here that i am still figuring out**: when you page down after a fresh reload, then search something, then clear your search and scroll all the way down to see all the results - one page will be missing. Now re-sort the view and scroll back down, all pages are present. There's a rogue paint to the grid that shouldn't be happening that I've yet to trace.~ 

## components/infiniteScroller/index
- Also here I want to call out the changes I made. Turns out I was only about half way there with my firefox + iframe + intersection observer. The layout effect hook was getting called repeatedly because the refs were updating with the callback from what I can see. So checking for the `currentIntersectingRef` before observing fixes it. The other thing here is that we're ONLY looking at .observe and .unobserve to trigger the effect - if you switch this to just `observer` it will break. So I disabled the exhaustive deps rule. 
- Simplified the surrounding message logic 

Todo
- [x] fix that bug I mentioned above 
- [ ] UI for infinite scroll messaging and scroll to top (will be separate) 
- [ ] I want opinions on separating out the API related constants but also that might be good to do separately 

## Testing 
- Pull the branch and run locally. Scroll through your stories. For this to work you should have more than a page's worth, which is 10 right now. If you want to you can update `ITEMS_PER_PAGE` to something smaller, but it's still easier to test this if you have at least 3 pages to go through. 
- So scroll through stuff
- Re-sort, see that the stories re arrange and are still scrolling the right amount so that you can see all of them 
- click on a filter and see that it updates 
- search for something 

![smalldemo](https://user-images.githubusercontent.com/10720454/79807270-40a4fa00-831f-11ea-9171-e6b69dd41bd8.gif)
